Refactor API re-authentication logic

cead22 opened the pull request:
<Assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit.>

<Explanation of the change or anything fishy that is going on>

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/141698

### Tests
**Reauthentication**

I applied this diff on chat to see the flow clearly in the console 
```diff
diff --git a/src/libs/API.js b/src/libs/API.js
index 5b955ed6..800318a5 100644
--- a/src/libs/API.js
+++ b/src/libs/API.js
@@ -104,9 +104,11 @@ function request(command, parameters, type = 'post') {
                 // Pause network requests and then queue the original one with Network.post,
                 // that way once we're done re-authenticating the original request will fire off again
                 Network.pauseRequestQueue();
+                console.debug('keke queueing', command, JSON.stringify(parameters));
                 Network.post(command, parameters, type);
 
                 // Trigger re-authentication
+                console.debug('keke auth token expired');
                 Onyx.set(ONYXKEYS.REAUTHENTICATING, true);
             }
 
diff --git a/src/libs/Network.js b/src/libs/Network.js
index 0f569587..6e7ad3d5 100644
--- a/src/libs/Network.js
+++ b/src/libs/Network.js
@@ -50,6 +50,8 @@ function processNetworkRequestQueue() {
             return;
         }
 
+        console.debug('keke firing off', queuedRequest.command, JSON.stringify(queuedRequest.data));
+
         const finalParameters = _.isFunction(enhanceParameters)
             ? enhanceParameters(queuedRequest.command, queuedRequest.data)
             : queuedRequest.data;
@@ -93,6 +95,7 @@ function post(command, data, type) {
  * Prevent the network queue from being processed
  */
 function pauseRequestQueue() {
+    console.debug('keke pausing network request');
     isQueuePaused = true;
 }
 
@@ -100,6 +103,7 @@ function pauseRequestQueue() {
  * Allow the network queue to continue to be processed
  */
 function unpauseRequestQueue() {
+    console.debug('keke unpausing network request');
     isQueuePaused = false;
 }
 
diff --git a/src/libs/actions/Session.js b/src/libs/actions/Session.js
index 21be7495..2f9800d6 100644
--- a/src/libs/actions/Session.js
+++ b/src/libs/actions/Session.js
@@ -24,6 +24,7 @@ let creatingLogin = false;
  * Refreshes the authToken by calling Authenticate with the stored credentials
  */
 function reauthenticate() {
+    console.debug('keke reauthenticating');
     API.Authenticate({
         useExpensifyLogin: false,
         partnerName: CONFIG.EXPENSIFY.PARTNER_NAME,
@@ -42,6 +43,7 @@ function reauthenticate() {
 
             // The authentication process is finished so the network can be unpaused to continue
             // processing requests
+            console.debug('keke unpausing network request');
             Network.unpauseRequestQueue();
         })
         .catch((error) => {

```

I also made auth tokens expire in 15 seconds with this diff

```diff
diff --git a/auth/lib/AuthToken.cpp b/auth/lib/AuthToken.cpp
index c1b648ebf..12cc1827b 100644
--- a/auth/lib/AuthToken.cpp
+++ b/auth/lib/AuthToken.cpp
@@ -918,7 +918,7 @@ string AuthToken::create(int64_t accountID, int64_t partnerID, const bool isVali
     authToken["accountID"] = to_string(accountID);
     authToken["partnerID"] = to_string(partnerID);
     authToken["issued"] = to_string(STimeNow());
-    authToken["expires"] = to_string(STimeNow() + expires);
+    authToken["expires"] = to_string(STimeNow() + 15 * STIME_US_PER_S);
     authToken["type"] = typeStrings[type];
 
     if (requestingAccountID != 0) {

```

- I signed out
- Signed back in
- Sent a message
- Waited 15 seconds
- Sent another message
- Saw the following in the console (ignore the `keke`, that was for easy filtering in the console)
```
Network.js:53     keke firing off Report_SetLastReadActionID {"accountID":1,"reportID":13,"sequenceNumber":43}
Network.js:53     keke firing off Report_AddComment {"reportID":13,"reportComment":"now?"}
Network.js:53     keke firing off Report_SetLastReadActionID {"accountID":1,"reportID":13,"sequenceNumber":44}
Network.js:98     keke pausing network request
API.js:107        keke queueing Report_AddComment {"reportID":13,"reportComment":"now?"}
API.js:111        keke auth token expired
Session.js:27     keke reauthenticating
Network.js:53     keke firing off Authenticate {"useExpensifyLogin":false,"partnerName":"android","partnerPassword":"c3a9ac418ea3f152aae2","partnerUserID":"react  native  chat  389b2836  36b0  cc46  385e  dcc66e1d4178","partnerUserSecret":"d7dd8d55  ad9a  9af0  a2d1  f92393bab4cf","doNotRetry":true,"forceNetworkRequest":true}
Network.js:98     keke pausing network request
API.js:107        keke queueing Report_SetLastReadActionID {"accountID":1,"reportID":13,"sequenceNumber":44}
API.js:111        keke auth token expired
Session.js:46     keke unpausing network request
Network.js:106    keke unpausing network request
Network.js:53     keke firing off Report_SetLastReadActionID {"accountID":1,"reportID":13,"sequenceNumber":44}
```

### Screenshots
<Add any necessary screenshots for all platforms if your change added or updated UI>

cead22 committed:
Add reauthenticate to Session

cead22 committed:
Add Authenticator lib

cead22 committed:
Init Authenticator when initing app

cead22 committed:
Add Credentials actions

cead22 committed:
Fix code according with latest libs

cead22 committed:
Remove dupe logic from Session

cead22 committed:
Add Onyx key for REAUTHENTICATING

cead22 committed:
Refactor API logic

tgolen reviewed (commented):
Hey Carlos, here are some thoughts on the code and a suggestion on an alternative approach that might make the re-authentication flow a little easier to follow.

cead22 committed:
Remove silly function

cead22 committed:
Refactor with simpler design

cead22 committed:
Update comment

cead22 committed:
Remove bad (and dupe) call to Network.post

cead22 committed:
Revert back to simpler ONYXKEYS.REAUTHENTICATING

cead22 committed:
Update var name since Ion is old school

cead22 committed:
Move reauthenticate call to session

cead22 committed:
Remove Authenticator

cead22 committed:
Consolidate authentication bits into Session

cead22 committed:
Reorganize file

Before:
- Let
- Onyx.connect
- Function
- Let
- Onyx.connect
- Function
- Let
- Onyx.connect
- Function

After:
- Let
- Let
- Let
- Function
- Function
- Function
- Onyx.connect
- Onyx.connect
- Onyx.connect

cead22 committed:
Remove unnecessary code

cead22 commented on src/libs/actions/Session.js:
```diff
@@ -3,16 +3,53 @@ import Str from 'expensify-common/lib/str';
+let isReauthenticating = false;
+
+// Used to prevent calling CreateLogin more than once since this callback is triggered when we set
+// authToken, loading, error, etc
```
> Hm, I see. It feels to me like we're maybe putting too many things inside the `session` Onyx object then

Yup



> For example, in the instance where `INSIDE createLogin true` then shouldn't we just prevent `createLogin()` from being called a second time?

We are with
```js
     if (creatingLogin) {
         return;
     }
```

> What if `session.loading` and `session.error` were moved to a different Onyx key?

We can, do you think that should be part of this PR? I can do it but it'll delay the PR, just like addressing [this comment](https://github.com/Expensify/Expensify.cash/pull/915#discussion_r541080220), both of which aren't part of the original scope, and since I've been working on this refactor for a while now, I'd prefer if we don't keep adding scope to this change. What do you think? I you agree I'll create issue for both of those things

tgolen replied:
> We are with

That's being done from inside of `createLogin()` which is not what I meant. I meant to not call `createLogin()` in the first place if we don't need to.

As it turns out, a lot of this is a non-issue and I'm already having to change a lot of this flow in https://github.com/Expensify/Expensify.cash/pull/1025 (that's the bad part about your PR here still being a WIP). I'm removing `session.loading` but for now I am leaving `session.error`.

tgolen commented:
I'm just going to close this since it's not a priority.
